### ISSUE TITLE
What remains to be done till the next stable release?

### ISSUE THREAD

**Issue by JBKahn** [schematics/schematics#398]
I wasn't sure of a better place to ask but what remains to be done till the next stable release?

**Comment by sheilatron** [comment]
Hi @bintoro do you need any volunteer contributions to help get 2.0dev up to an alpha or beta release quality? So far I have not seen any issues in my limited use...

**Comment by bintoro** [comment]
Hey @JBKahn, @sheilatron et al.

Sorry about the lack of response — I've been super busy the past couple of weeks.

There are not *too* many things left to be done before 2.0 could be considered a release candidate. In any case, as you've noticed, the development version is in a working condition, even if some APIs might not be final.

---

Off the top of my head, perhaps the most important task is finalizing the new exception design and deciding what we want error messages to look like.

Historically, the messages are just strings, and even though you can provide your own format strings, the mechanism doesn't lend itself well to localization, for instance. Also, the formats have no standard structure — what `{0}` means is completely up to the type.

It's pretty clear that we should be constructing the messages in a more structured way that makes is easy to programmatically determine what went wrong. The new exceptions have laid the groundwork for this: the message objects are now a custom class instead of `str`, but I still haven't thought through how best to exploit that.

As always, it's not a hugely complex issue per se, but backward-compatibility considerations make it more so.

For example, I'm not sure the existing ability to provide your own message strings to the field instance is very useful. All I need is

* a basic human-readable message for **developers** to look at
* an ID such as `"max_value"` that indicates in a standardized way the constraint that was violated.

The program can then come up with the actual **user-facing** error message however it likes.

I'd love to hear how people are currently going about this in practice and what kind of solution would work best for them.

---

I'll go over my notes and post a summary of the remaining items sometime this week.

Let's keep this issue open to discuss the 2.0 release and track progress!

**Comment by JBKahn** [comment]
The way I've been using schematics is with three additional layers:
1) A few overrides and super calls to allow me to pass in a Django model (definitely not something I would ever expect to change about the library being agnostic of franework) where to_native can handle models and foreign fields. Nothing major.
2) The one that is debatably good is how we chose to write our own validation exceptions on top of schematics. We came up with our own error format that we liked which was:
```
{'__all__': ['blah'], 'field1': error1, 'field2': error2, 'field3':[{error_a}, {}, {arrorb}]}
```
as a python dict so that we can easily map errors to fields they came from either by key, position in a list, or key in a dictionary.
```
And we return that dictionary to the consumer.

3) We also took the liberty of wrappiing the framework in a class so that we can more easily validate the whole thing without having to know the order of items in the validator.

**Comment by bintoro** [comment]
I've been analyzing the outstanding issues, and in fact we're very close to be able to call this 2.0. There are several things in the pipeline that would be nice to have, but only a few represent truly incompatible changes.

* #375 discusses changes to the Model API that need to be decided before release. I think the outcome of the main discussion there is that we could clean up the Model methods a little bit and provide a mixin that implements full `MutableMapping` functionality. (I believe methods like `items()` are rarely used.)

* #372 introduced the notion of unset, or undefined, fields that don't have a value. (This is to be able to distinguish `None` from missing.) The feature entails pretty significant changes that have been hashed out here to an extent, but very few users have actually commented on anything. Mostly it's simply additional functionality, but the new behavior of default values is a notable breaking change (to the better, I hope).

Previously I mentioned error handling. I think I've now thought it through, and it seems that, unless there's a reason to fundamentally change the error message dictionary structure, implementing formal error IDs to accompany the current human-readable strings is a fairly straightforward additive change and hence not a blocker.

@JBKahn I'm curious, how does the custom error dictionary differ from the standard one?

Schematics v1 does have a known issue with `ListType` errors: there's no index to the offending item. But this is fixed in v2; there's an example at the end of #369.

**Comment by lkraider** [comment]
2.0.0 now on PyPI:
https://pypi.python.org/pypi/schematics/2.0.0
